A29-xds-tls-security.md

```diff
@@ -0,0 +1,411 @@
+of the received `UpstreamTlsContext` then gRPC validates the SAN entries in the server certificate
+by matching the `match_subject_alt_names` values using the
+[StringMatcher](https://github.com/envoyproxy/envoy/blob/main/api/envoy/type/matcher/string.proto#L20)
+semantics. This is called server authorization because this is how a client "authorizes" a server for
+the connection.
+
+The implementation is language dependent. As an example, Java uses an implementation
+of `X509TrustManager` through the `SslContext` provided to the client sub-channel. The other
+languages use similar hooks provided by the underlying TLS framework to implement server
+authorization.
+
+#### Server Side Flow
+
+The gRPC server xDS flow is described in
+[gRPC Server ](./A27-xds-global-load-balancing.md#grpc-client-architecture). The server
+configuration comes from the
+[`Listener`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/listener/v3/listener.proto#L39)
+resource which contains a list of
+[`FilterChain`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/listener/v3/listener_components.proto#L193)s.
+The `FilterChain` contains the `DownstreamTlsContext` that is applied to an incoming connection
+after selecting the best-matching `FilterChain` as described in
+[FilterChainMatch](A36-xds-for-servers.md#filterchainmatch). If the server is using `XdsServerCredentials`,
+gRPC processes the `DownstreamTlsContext` as described [above](#lds-for-server-side-security) and then
+uses the flow described above in
+[Use of Bootstrap File by the Certificate Provider Plugin Framework](#use-of-bootstrap-file-by-the-certificate-provider-plugin-framework)
+to fetch the requisite `CertificateProvider`s (one or two as the case may be). These `CertificateProvider`s,
+in one form or another, need to be made available to the incoming connection. Similar to the
+client side, this part of the flow is language dependent. For example, Java constructs a dynamic
+`SslContextProvider` that is directly usable by the connection's TLS handshaker to build an
+`SslContext` for the impending handshake.
+
+## Rationale
+
+gRPC has been steadily adding support for various xDS features including load balancing and
+advanced traffic management. xDS-based security was next in the roadmap and this proposal
+addresses that. gRPC makes a proxyless service mesh (PSM) possible. A secure proxyless
+service mesh requires the security features described in this proposal.
+
+The Certificate Provider Plugin Framework does not depend on an agent (SDS server). The
+SDS server/agent based solution compromised security (by sharing the private key with the
+agent). The combination of Envoy sidecar and the SDS server was also somewhat tied to the
+Kubernetes architecture where Envoy runs in a sidecar container and the SDS server runs
+as a Node agent in the cluster. The Certificate Provider Plugin Framework has eliminated that
```

Comment:
I wouldn't say "eliminated that dependency" as the plugin framework didn't do that at all; swapping to file-based cert did that and it could be done outside of the plugin framework. The plugin framework allows different environments to have different dependencies and it allows the control plane to be agnostic to the specific environment the client is running (since generally the control plane has either hard-coded paths assuming a single environment or requires the client to specify values in metadata and have it echoed back).

Reply:
fixed. PTAL